File: src/types/artifacts.ts
export interface ModelConfig {
  input_dim: number;
  hidden_dims: number[];
  dropout: number;
  best_threshold: number;
  seed?: number;
  csv_path?: string;
}

export interface HistoryRow {
  epoch: number;
  train_loss: number;
  train_accuracy: number;
  train_f1: number;
  train_roc_auc: number;
  val_accuracy: number;
  val_f1: number;
  val_roc_auc: number;
  lr: number;
}

export interface PredictionRow {
  sampleIndex: number;
  true_label: number;
  pred_probability: number;
  pred_label: number;
}

export interface ThresholdRow {
  threshold: number;
  tp: number;
  tn: number;
  fp: number;
  fn: number;
  precision: number;
  recall: number;
  f1: number;
  balanced_accuracy: number;
}

export interface SimulatedFeatureRow {
  sampleIndex: number;
  values: number[];
}

export interface ArtifactRecord {
  name: string;
  path: string;
  available: boolean;
  kind: "json" | "csv" | "image" | "binary" | "unknown";
}

export interface ArtifactBundle {
  config: ModelConfig;
  history: HistoryRow[];
  predictions: PredictionRow[];
  thresholds: ThresholdRow[];
  simulatedFeatures: SimulatedFeatureRow[];
  artifacts: ArtifactRecord[];
  summary: {
    historyCount: number;
    predictionCount: number;
    positiveCount: number;
    negativeCount: number;
    averageProbability: number;
    threshold: number;
  };
  sourceMode: "local" | "backend" | "fallback";
  warnings: string[];
}

File: backend/app/schemas/artifacts.py
from pydantic import BaseModel


class ConfigResponse(BaseModel):
    available: bool
    data: dict | None
    source: str


class HistoryRow(BaseModel):
    epoch: int
    train_loss: float
    train_accuracy: float
    train_f1: float
    train_roc_auc: float
    val_accuracy: float
    val_f1: float
    val_roc_auc: float
    lr: float


class PredictionRow(BaseModel):
    true_label: int
    pred_probability: float
    pred_label: int


class HistoryResponse(BaseModel):
    available: bool
    count: int
    data: list[HistoryRow]
    source: str


class PredictionResponse(BaseModel):
    available: bool
    count: int
    data: list[PredictionRow]
    source: str


class ThresholdRow(BaseModel):
    threshold: float
    tp: int
    tn: int
    fp: int
    fn: int
    precision: float
    recall: float
    f1: float
    balanced_accuracy: float


class ThresholdResponse(BaseModel):
    available: bool
    count: int
    data: list[ThresholdRow]
    source: str
